core/src/main/scala/org/bitcoins/core/crypto/TxSigComponent.scala

```diff
@@ -166,6 +186,24 @@ object BaseTxSigComponent {
 
 }
 
+object P2SHTxSigComponent {
+
+  private case class P2SHTxSigComponentImpl(
+      transaction: Transaction,
+      inputIndex: UInt32,
+      output: TransactionOutput,
+      flags: Seq[ScriptFlag])
+      extends P2SHTxSigComponent
```

Comment:
I think I'd prefer to have the top level class be a `case class` rather than using the `impl` pattern as it makes things nicer to pattern match on

Reply:
In the past this pattern has been used in place of `require` statements to restrict what constructions are legal but you have invariants stated in the super class so using the `impl` pattern is unnecessary